src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Linq;
+using System.Text;
+using CSharpFunctionalExtensions;
+using NBitcoin;
+using Stratis.SmartContracts.Core.State;
+
+namespace Stratis.Features.SystemContracts.Contracts
+{
+    /// <summary>
+    /// Sample auth contract.
+    /// </summary>
+    public class AuthContract
+    {
+        public AuthContract(IStateRepositoryRoot state, ISystemContractContainer systemContractContainer)
+        {
+            this.State = state;
+            this.SystemContractContainer = systemContractContainer;
+
+            if (!this.Initialized)
+            {
+                this.Initialized = true;
+            }
+        }
+
+        public bool Initialized
+        {
+            get
+            {
+                // We're doing this the hard way for demonstration purposes,
+                // but we could inject the SC PersistentState here to avoid this if we wanted.
+                var data = this.State.GetStorageValue(Identifier.Data, Encoding.UTF8.GetBytes("Initialized"));
+                return data == null ? false : BitConverter.ToBoolean(data);
+            }
+
+            set
+            {
+                this.State.SetStorageValue(Identifier.Data, Encoding.UTF8.GetBytes("Initialized"), BitConverter.GetBytes(value));
+            }
+        }
+
+        public static Identifier Identifier => new Identifier(new uint160(SCL.Crypto.SHA3.Keccak256(Encoding.UTF8.GetBytes(nameof(AuthContract))).Take(20).ToArray()));
+
+        public IStateRepositoryRoot State { get; }
+
+        public ISystemContractContainer SystemContractContainer { get; }
+
+        public Network Network { get; }
+
+        public bool IsAuthorised(string[] signatures)
+        {
+            return VerifySignatories(signatures, this.SystemContractContainer.PrimaryAuthenticators?.Signatories ?? new string[] { });
+        }
+
+        private bool VerifySignatories(string[] signatures, string[] signatories)
+        {
+            if(signatures.Length == 0)
+                return false;
+
+            return signatures[0] == "secret";
+        }
+
+        /// <summary>
+        /// Dynamic method resolution and dependency injection for the <see cref="AuthContract"/> type
+        /// to allow it to be invoked on-chain.
+        /// 
+        /// Why is this necessary?
+        /// For system contracts, call data is received from a transaction and deserialized to strings.
```

Comment:
This should be true for both embedded contracts and on-chain contracts. We should standardize on a common approach for both. I.e. "normal" smart contracts should also be able to use dispatchers so that we reach a common code path sooner.

Reply:
We could but that's a substantial change to smart contracts and outside the scope of this work.

Reply:
On the contrary, the idea is to reduce change by being very selective and minimal with what is changed.

Reply:
It does the exact opposite though, by adding new functionality to Cirrus.